File: mars-rover/java/src/main/java/mars/rover/Rover.java
package mars.rover;

import java.util.Arrays;
import java.util.List;
import lombok.*;

@EqualsAndHashCode
@ToString
public class Rover {

    public static enum Orientation {
        N, S, W, E
    };

    public static enum Instruction {
        Left, Right, Move
    };


    public boolean someLibraryMethod() {
        return false;
    }

    private Integer x;
    private Integer y;
    public Orientation orientation;

    public Rover(Integer x, Integer y, Orientation o) {
        this.x = x;
        this.y = y;
        this.orientation = o;
    }

    public Rover move() {
        switch (orientation) {
        case N:
            return new Rover(x, y + 1, orientation);
        case S:
            return new Rover(x, y - 1, orientation);
        case E:
            return new Rover(x + 1, y, orientation);
        default:
        case W:
            return new Rover(x - 1, y, orientation);
        }
    }

    public Rover exec(List<Instruction> is) {
        Rover r = this;

        for (Instruction i : is)
            r = r.exec(i);

        return r;
    }

    public Rover exec(Instruction i) {
        switch (i) {
        case Move:
            return this.move();
        case Left:
            return this.rotateLeft();
        default:
        case Right:
            return this.rotateRight();
        }
    }

    public Rover rotateRight() {
        return rotateLeft().rotateLeft().rotateLeft();
    }

    public Rover rotateLeft() {
        switch (this.orientation) {
        case N:
            return new Rover(x, y, Orientation.W);
        case W:
            return new Rover(x, y, Orientation.S);
        case S:
            return new Rover(x, y, Orientation.E);
        case E:
            return new Rover(x, y, Orientation.N);
        default:
            return this;
        }
    }
}

File: mars-rover/kotlin/src/main/kotlin/mars/rover/Rover.kt
package mars.rover

data class Pos(val x: Int, val y: Int)

enum class Instruction { Move, Left, Right }

enum class Dir { N, S, E, W }

data class Rover(val pos: Pos, val dir: Dir)

fun Rover.rotateLeft(): Rover = when (this.dir) {
    Dir.N -> this.copy(dir = Dir.W)
    Dir.S -> this.copy(dir = Dir.E)
    Dir.E -> this.copy(dir = Dir.N)
    Dir.W -> this.copy(dir = Dir.S)
}

fun Rover.rotateRight(): Rover = when (this.dir) {
    Dir.N -> this.copy(dir = Dir.E)
    Dir.S -> this.copy(dir = Dir.W)
    Dir.E -> this.copy(dir = Dir.S)
    Dir.W -> this.copy(dir = Dir.N)
}

fun Rover.advance(): Rover = when (this.dir) {
    Dir.N -> this.copy(pos = this.pos.copy(y = this.pos.y + 1))
    Dir.S -> this.copy(pos = this.pos.copy(y = this.pos.y - 1))
    Dir.E -> this.copy(pos = this.pos.copy(x = this.pos.x + 1))
    Dir.W -> this.copy(pos = this.pos.copy(x = this.pos.x - 1))
}

fun Rover.exec(i: Instruction): Rover = when (i) {
    Instruction.Left -> this.rotateLeft()
    Instruction.Right -> this.rotateRight()
    Instruction.Move -> this.advance()
}

fun Rover.exec(insts: List<Instruction>): Rover = insts.fold(this, {
    r: Rover, i: Instruction -> r.exec(i)
})
